src/main/java/jenkins/plugins/openstack/compute/JCloudsPreCreationThread.java

```diff
@@ -0,0 +1,92 @@
+import hudson.model.AsyncPeriodicWork;
+
+/**
+ * Periodically ensure enough slaves are created.
+ */
+@Extension @Restricted(NoExternalUse.class)
+public final class JCloudsPreCreationThread extends AsyncPeriodicWork {
+    private static final Logger LOGGER = Logger.getLogger(JCloudsPreCreationThread.class.getName());
+
+    public JCloudsPreCreationThread() {
+        super("OpenStack slave pre-creation");
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return MIN * 2;
+    }
+
+    @Override
+    public void execute(TaskListener listener) {
+        for (JCloudsCloud cloud : JCloudsCloud.getClouds()) {
+            for (JCloudsSlaveTemplate template : cloud.getTemplates()) {
+                SlaveOptions slaveOptions = template.getEffectiveSlaveOptions();
+                int instancesMin = slaveOptions.getInstancesMin();
+                int retentionTime = slaveOptions.getRetentionTime();
+                if (instancesMin > 0) {
+                    int globalMaxInstances = cloud.getEffectiveSlaveOptions().getInstanceCap();
+                    int templateMaxInstances = slaveOptions.getInstanceCap();
+                    int maxNodes = Math.min(templateMaxInstances, globalMaxInstances);
+                    // If retentionTime==0, take this as an indication that "used" instances should not
+                    // be reused and thus do not count them as running instances.
+                    int runningNodeTotal = template.getActiveNodesTotal(retentionTime == 0);
+                    int desiredNewInstances = Math.min(instancesMin, maxNodes) - runningNodeTotal;
+                    if (desiredNewInstances > 0) {
+                        LOGGER.log(Level.INFO, "Pre-creating " + desiredNewInstances + " instance(s) for template " + template.name + " in cloud " + cloud.name);
+                        for (int i = 0; i < desiredNewInstances; i++) {
+                            try {
+                                cloud.doProvisionSlave(template);
+                            } catch (Throwable ex) {
+                                LOGGER.log(Level.SEVERE, "Failed to pre-create instance from template " + template.name, ex);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+    * Should a slave be retained to meet the minimum instances constraint?
+    */
+    public static boolean shouldSlaveBeRetained(JCloudsSlave slave) {
+        String templateName = slave.getId().getTemplateName();
+        String cloudName = slave.getId().getCloudName();
+        if (templateName != null && cloudName != null) {
+            JCloudsCloud cloud = JCloudsCloud.getByName(cloudName);
+            if (cloud != null) {
+                JCloudsSlaveTemplate template = cloud.getTemplate(templateName);
+                if (template != null) {
+                    SlaveOptions slaveOptions = template.getEffectiveSlaveOptions();
```

Comment:
Here it would be more safe and simpler to extract the slave options from slave directly to avoid the chance the cloud or template got renamed or there is no template associated.

Reply:
I disagree: I believe the safest and the simplest thing to do here (and this is what I've implemented thus far) is to only consider the setting (`instancesMin`) on the *template*; if a node has no template, then it's not part of the "pre-provisioning" program anymore.  Not doing this leads to weird behaviour where, when a node has `instancesMin=5` and the related template has a different value, say, 3, Jenkins keeps 5 instances and it's not really clear why.

All in all, I suggest we simply always use the `instancesMin` settings from the related template and if the related template doesn't exist, we act as if `instancesMin` is zero.

Reply:
Ok, makes sense to me.